query/src/main/java/org/infinispan/query/impl/LifecycleManager.java

```diff
@@ -311,25 +305,23 @@ private boolean verifyChainContainsQueryInterceptor(ComponentRegistry cr) {
       return interceptorChain != null && interceptorChain.containsInterceptorType(QueryInterceptor.class, true);
    }
 
-   private SearchIntegrator getSearchFactory(String cacheName, IndexingConfiguration indexingConfiguration, ComponentRegistry cr) {
-      Object component = cr.getComponent(SearchIntegrator.class);
-      SearchIntegrator searchFactory = null;
-      if (component instanceof SearchIntegrator) { //could be the placeholder Object REMOVED_REGISTRY_COMPONENT
-         searchFactory = (SearchIntegrator) component;
-      }
-      //defend against multiple initialization:
-      if (searchFactory == null) {
-         ClassLoader aggregatedClassLoader = makeAggregatedClassLoader(cr.getGlobalComponentRegistry().getGlobalConfiguration().classLoader());
-         Collection<ProgrammaticSearchMappingProvider> programmaticSearchMappingProviders = new HashSet<>(ServiceFinder.load(ProgrammaticSearchMappingProvider.class, aggregatedClassLoader));
-         programmaticSearchMappingProviders.add(new DefaultSearchMappingProvider());
-         Collection<LuceneAnalysisDefinitionProvider> analyzerDefProviders = ServiceFinder.load(LuceneAnalysisDefinitionProvider.class, aggregatedClassLoader);
-         // Set up the search factory for Hibernate Search first.
-         SearchConfiguration config = new SearchableCacheConfiguration(indexingConfiguration.indexedEntities(),
-               indexingConfiguration.properties(), programmaticSearchMappingProviders, analyzerDefProviders, cr, aggregatedClassLoader);
-
-         searchFactory = new SearchIntegratorBuilder().configuration(config).buildSearchIntegrator();
-         cr.registerComponent(searchFactory, SearchIntegrator.class);
+   private SearchIntegrator createSearchIntegrator(IndexingConfiguration indexingConfiguration, ComponentRegistry cr, ClassLoader aggregatedClassLoader) {
+      // prevent multiple init
```

Comment:
This shouldn't be necessary either, `cacheStarting()` only runs once for a cache.

I think this pattern first appeared when query added the interceptors to the current chain *and* to the interceptors configuration, meaning the interceptor would already be registered after a restart. But even in that case it would be simpler if we always registered the interceptors the same way, without modifying the configuration.

Reply:
Nope. That pattern appeared because it registers the SearchIntegrator on starting but has no way to unregister it on stopping. But it does ```close``` the SearchIntegrator. After restart it needs a way to detect that SearchIntegrator it find in CR is garbage and throw it away, so a dummy Object is used to replace an existing component, as marker for a stopped SearchIntegrator.  This dirty trick is no longer needed because SearchIntegrator.isStopped was introduced at some later point in time.

The ```// prevent multiple init``` comment is there also because this method previously called ```getSearchFactory``` used to be called multiple times in older code. Now it's only one invocation and I have renamed it. I'll remove the comment.

Reply:
Actually, the whole thing boils down to the fact ```getSearchFactory``` is no longer invoked multiple times, so no danger of double init appears. All that code will be gone, no just the comment. I'm so happy we remove that crap.

Reply:
actually, I think I'll keep the ```throw new IllegalStateException("SearchIntegrator already initialized!");``` for a while. I don't trust anything and anybody these days :)